File: test_files/vulnerable.java
/**
 * Test file with various Java vulnerabilities
 */

import java.io.*;
import java.sql.*;
import java.util.Random;
import java.security.MessageDigest;
import javax.servlet.http.HttpServletRequest;

public class VulnerableJavaCode {
    
    // Hard-coded credentials - CWE-798
    private static final String PASSWORD = "admin123";
    private static final String API_KEY = "sk-1234567890abcdef";
    private static final String DB_URL = "jdbc:mysql://localhost:3306/db?user=admin&password=secret";
    
    // SQL Injection vulnerability - CWE-89
    public ResultSet getUserData(String userId) throws SQLException {
        Connection conn = DriverManager.getConnection(DB_URL);
        Statement stmt = conn.createStatement();
        
        // Vulnerable: string concatenation in SQL
        String sql = "SELECT * FROM users WHERE id = " + userId;
        return stmt.executeQuery(sql);  // CWE-89: SQL Injection
    }
    
    // Command Injection vulnerability - CWE-78
    public void executeCommand(String userInput) throws IOException {
        // Vulnerable: direct command execution
        Runtime.getRuntime().exec("ping " + userInput);  // CWE-78: Command Injection
        
        // Also vulnerable through ProcessBuilder
        ProcessBuilder pb = new ProcessBuilder("sh", "-c", "cat " + userInput);
        pb.start();  // CWE-78: Command Injection
    }
    
    // Buffer overflow potential - CWE-787 (simulated in Java context)
    public void unsafeBufferOperation(String input) {
        // Vulnerable: no bounds checking
        char[] buffer = new char[256];
        for (int i = 0; i < input.length(); i++) {
            buffer[i] = input.charAt(i);  // CWE-787: Potential buffer overflow
        }
    }
    
    // Path Traversal vulnerability - CWE-22
    public String readFile(String filename) throws IOException {
        // Vulnerable: no path validation
        FileReader fileReader = new FileReader(filename);  // CWE-22: Path Traversal
        BufferedReader bufferedReader = new BufferedReader(fileReader);
        
        StringBuilder content = new StringBuilder();
        String line;
        while ((line = bufferedReader.readLine()) != null) {
            content.append(line).append("\n");
        }
        
        bufferedReader.close();
        return content.toString();
    }
    
    // Weak cryptography - CWE-327
    public String weakHash(String input) throws Exception {
        // Vulnerable: using MD5
        MessageDigest md5 = MessageDigest.getInstance("MD5");
        byte[] hash = md5.digest(input.getBytes());
        
        // Vulnerable: using SHA1
        MessageDigest sha1 = MessageDigest.getInstance("SHA-1");
        byte[] hash2 = sha1.digest(input.getBytes());
        
        return bytesToHex(hash);  // CWE-327: Weak Hash Algorithm
    }
    
    // XSS vulnerability simulation - CWE-79
    public String generateHtml(HttpServletRequest request) {
        String userInput = request.getParameter("comment");
        
        // Vulnerable: direct HTML output without encoding
        return "<div>" + userInput + "</div>";  // CWE-79: XSS
    }
    
    // Improper Input Validation - CWE-20
    public int processAge(String ageString) {
        // Vulnerable: no validation
        return Integer.parseInt(ageString);  // CWE-20: Could throw exception
    }
    
    // Weak random number generation - CWE-330
    public String generateSessionId() {
        Random random = new Random();
        
        // Vulnerable: using weak random for security
        return String.valueOf(random.nextInt(10000));  // CWE-330: Weak Random
    }
    
    // Use After Free simulation - CWE-416 (not directly applicable in Java, but similar pattern)
    public void useAfterClose() throws IOException {
        FileInputStream fis = new FileInputStream("test.txt");
        fis.close();
        
        // Vulnerable: using closed resource
        fis.read();  // CWE-416: Use After Free (equivalent in Java)
    }
    
    // Improper Authentication - CWE-287
    public boolean authenticateUser(String username, String password) {
        // Vulnerable: hardcoded authentication
        if (username.equals("admin") && password.equals("admin123")) {
            return true;  // CWE-287: Improper Authentication
        }
        return false;
    }
    
    // Missing Authorization - CWE-862
    public void deleteUser(String userId) {
        // Vulnerable: no authorization check
        // Should check if current user has permission to delete
        System.out.println("Deleting user: " + userId);  // CWE-862: Missing Authorization
    }
    
    // Utility method
    private String bytesToHex(byte[] bytes) {
        StringBuilder result = new StringBuilder();
        for (byte b : bytes) {
            result.append(String.format("%02x", b));
        }
        return result.toString();
    }
    
    public static void main(String[] args) {
        VulnerableJavaCode vulnerable = new VulnerableJavaCode();
        
        try {
            // Test various vulnerability patterns
            vulnerable.getUserData("1");
            vulnerable.executeCommand("localhost");
            vulnerable.unsafeBufferOperation("test input");
            vulnerable.readFile("test.txt");
            vulnerable.weakHash("password");
            vulnerable.processAge("25");
            vulnerable.generateSessionId();
            vulnerable.authenticateUser("admin", "admin123");
            vulnerable.deleteUser("123");
            
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
        }
    }
}

File: test_suite/unit/fixtures/vulnerable.py
#!/usr/bin/env python3
"""
Test file with various vulnerabilities to demonstrate performance improvements.
"""

import os
import sqlite3
import subprocess
import hashlib

def sql_injection_vulnerability():
    """SQL injection vulnerability - CWE-89"""
    user_id = input("Enter user ID: ")
    # Vulnerable: direct string concatenation
    query = "SELECT * FROM users WHERE id = " + user_id
    
    conn = sqlite3.connect('database.db')
    cursor = conn.cursor()
    cursor.execute(query)  # CWE-89: SQL Injection
    return cursor.fetchall()

def command_injection_vulnerability():
    """Command injection vulnerability - CWE-78"""
    filename = input("Enter filename: ")
    # Vulnerable: direct command execution
    os.system("cat " + filename)  # CWE-78: OS Command Injection
    
    # Also vulnerable through subprocess
    subprocess.run("ping " + filename, shell=True)  # CWE-78

def hardcoded_credentials():
    """Hard-coded credentials - CWE-798"""
    password = "admin123"  # CWE-798: Hard-coded Password
    api_key = "sk-1234567890abcdef"  # CWE-798: Hard-coded API Key
    db_connection = "mongodb://admin:password@localhost:27017/db"  # CWE-798
    
    return password, api_key, db_connection

def weak_cryptography():
    """Weak cryptographic algorithms - CWE-327"""
    import hashlib
    
    # Vulnerable: using MD5
    weak_hash = hashlib.md5(b"password").hexdigest()  # CWE-327: Weak Hash
    
    # Vulnerable: using SHA1
    weak_hash2 = hashlib.sha1(b"password").hexdigest()  # CWE-327: Weak Hash
    
    return weak_hash, weak_hash2

def path_traversal_vulnerability():
    """Path traversal vulnerability - CWE-22"""
    filename = input("Enter filename: ")
    # Vulnerable: no path validation
    with open(filename, 'r') as f:  # CWE-22: Path Traversal
        return f.read()

def xss_vulnerability():
    """Cross-site scripting simulation - CWE-79"""
    user_input = input("Enter comment: ")
    # Vulnerable: direct HTML output without encoding
    html_output = f"<div>{user_input}</div>"  # CWE-79: XSS
    return html_output

def improper_input_validation():
    """Improper input validation - CWE-20"""
    user_age = input("Enter age: ")
    # Vulnerable: no validation
    age = int(user_age)  # CWE-20: Could crash on invalid input
    
    user_email = input("Enter email: ")
    # Vulnerable: no email validation
    return age, user_email

def use_of_eval():
    """Code injection through eval - CWE-94"""
    user_code = input("Enter Python expression: ")
    # Vulnerable: eval with user input
    result = eval(user_code)  # CWE-94: Code Injection
    return result

def weak_random_numbers():
    """Weak random number generation - CWE-330"""
    import random
    
    # Vulnerable: using weak random for security
    session_id = random.randint(1000, 9999)  # CWE-330: Weak Random
    csrf_token = str(random.random())  # CWE-330: Weak Random
    
    return session_id, csrf_token

def main():
    """Main function demonstrating various vulnerabilities"""
    print("Testing various vulnerability patterns...")
    
    try:
        # Test each vulnerability type
        sql_injection_vulnerability()
        command_injection_vulnerability()
        hardcoded_credentials()
        weak_cryptography()
        path_traversal_vulnerability()
        xss_vulnerability()
        improper_input_validation()
        use_of_eval()
        weak_random_numbers()
        
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()